src/test/java/de/tum/in/www1/artemis/service/connectors/LtiServiceTest.java

```diff
@@ -0,0 +1,286 @@
+        when(userService.createUser(username, groups, "", launchRequest.getLis_person_sourcedid(), launchRequest.getLis_person_contact_email_primary(), null, null, "en"))
+                .thenReturn(user);
+
+        onSuccessfulAuthenticationSetup(user, ltiUserId);
+
+        ltiService.handleLaunchRequest(launchRequest, exercise);
+
+        onSuccessfulAuthenticationAssertions(user, ltiUserId);
+        verify(userService).activateUser(user);
+
+        SecurityContextHolder.clearContext();
+        launchRequest.setContext_label("randomLabel");
+
+        InternalAuthenticationServiceException exception = assertThrows(InternalAuthenticationServiceException.class, () -> {
+            ltiService.handleLaunchRequest(launchRequest, exercise);
+        });
+
+        String expectedMessage = "User group not activated or unknown context_label sent in LTI Launch Request: " + launchRequest.toString();
+        assertThat(exception.getMessage().equals(expectedMessage));
+    }
+
+    @Test
+    public void handleLaunchRequest_noAuthenticationWasSuccessful() {
+        launchRequest.setCustom_require_existing_user(true);
+        when(ltiUserIdRepository.findByLtiUserId(launchRequest.getUser_id())).thenReturn(Optional.empty());
+        when(response.containsHeader("Set-Cookie")).thenReturn(true);
+        List<String> headers = new ArrayList<>();
+        headers.add("JSESSIONID=(123);");
+        when(response.getHeaders("Set-Cookie")).thenReturn(headers);
+        when(response.getHeader("Set-Cookie")).thenReturn(headers.get(0));
+        String sessionId = "(123)";
+
+        ltiService.handleLaunchRequest(launchRequest, exercise);
+
+        assertThat(ltiService.launchRequestForSession.containsKey(sessionId));
+        assertThat(ltiService.launchRequestForSession.containsValue(Pair.of(launchRequest, exercise)));
+        assertThat(ltiService.launchRequestForSession.get(sessionId).equals(Pair.of(launchRequest, exercise)));
+    }
+
+    @Test
+    public void onSuccessfulLtiAuthentication() {
+        ltiUserId.setLtiUserId("oldStudentId");
+        onSuccessfulAuthenticationSetup(user, ltiUserId);
+
+        ltiService.onSuccessfulLtiAuthentication(launchRequest, exercise);
+
+        onSuccessfulAuthenticationAssertions(user, ltiUserId);
+    }
+
+    private void onSuccessfulAuthenticationSetup(User user, LtiUserId ltiUserId) {
+        when(userService.getUserWithGroupsAndAuthorities()).thenReturn(user);
+        when(ltiUserIdRepository.findByUser(user)).thenReturn(Optional.of(ltiUserId));
+        when(ltiOutcomeUrlRepository.findByUserAndExercise(user, exercise)).thenReturn(Optional.of(ltiOutcomeUrl));
+    }
+
+    private void onSuccessfulAuthenticationAssertions(User user, LtiUserId ltiUserId) {
+        assertThat(user.getGroups().contains(courseStudentGroupName));
+        assertThat(ltiUserId.getLtiUserId().equals("ff30145d6884eeb2c1cef50298939383"));
```

Comment:
there are some Codcy issues, regarding this issue. Please fix those :) 

```suggestion
        assertThat("ff30145d6884eeb2c1cef50298939383".equals(ltiUserId.getLtiUserId()));
```